technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,122 @@
+```
+
+The CRD should also contain information that should be overridable such as the prefix, which is controlled by the namespace name. 
+
+## Medusa changes
+The tokenmap generated by Medusa will need to be augmented to provide the placement of nodes in the cluster: datacenter and rack.
+Currently Medusa is not dc/rack aware and it can probably be fairly surprising to users. It was built to backup a single DC per cluster and didn’t account for racks either.  
+Restoring from a DC/rack aware cluster to another one imposes some constraints on the source/target nodes mapping, which are currently impossible to deal with safely in a remote restore: 
+
+When using NetworkTopologyStrategy, the placement of replicas is dependent on the DC and rack layout.
+In this context, the replica for a token range will be placed on the node that owns the next token range in the next rack.
+We can't restore any node to any other node as we may end up putting several replicas in the same rack, and the nodes we'd restore to would possibly end up not containing data they're supposed be replica for.
+We can be rigid about the mapping and expect the same dc and rack names exclusively, or do a 1:1 mapping between DCs/Racks despite having different names (but expect the same number of both).
+
+
+## Restore Orchestration
+
+1. Get backup definition from storage backend
+![](img/K8ssandra-new-restore-design-1.jpg)  
+The restore operation will assume the CRDs have been synced with the remote storage backend. Backup topology (hostnames and tokens) will be retrieved from the new CRD format.
+
+
+2. Set the env variables on the restore containers
+![](img/K8ssandra-new-restore-design-2.jpg)  
+The backup definition will be used to map the backup hosts with the restore hosts and perform checks on whether or not the restore is possible:
+
+  - Same number of DCs
+  - Same number of nodes per DC
+  - Same rack distribution per DC
+  
+  Nodes will be mapped 1:1 between the source and destination topologies (sorted alphabetically). If the mapping fails to associate nodes 1:1, the restore operation must fail.
+
+  The init containers will get the following env variables set:
+
+  - Backup name
+  - Restore key
+  - Bucket prefix (new)
+  - In Place/Remote (new)
+
+  These values will then be used to invoke medusa.restore_node in the medusa-restore init container.
+The init-container needs to deal with mapping between source and dest topologies to self determine which backup it needs to restore. Pre-checks in the operator will guarantee that a 1:1 mapping is possible and topologies are compatible.
+
+3. Shutdown Cassandra to trigger the restore
+![](img/K8ssandra-new-restore-design-3.jpg)  
+
+  medusa-operator will patch the cassandradatacenter spec to provoke a shutdown of the Cassandra nodes through cass-operator. The pods will be restarted and the medusa-restore init container will trigger the restore.
```

Comment:
Is this going to happen all nodes at once, or this is going to be rolling restarts?

Reply:
All at once, so that we can guarantee a consistent restore from the apps perspective. In the case of in place restores, we're studying how we can perform live restores while staying consistent, but it'll be part of another task.